core/src/main/java/org/elasticsearch/cluster/metadata/IndexTombstone.java

```diff
@@ -0,0 +1,227 @@
+        if (this == other) {
+            return true;
+        }
+        if (other == null || getClass() != other.getClass()) {
+            return false;
+        }
+        @SuppressWarnings("unchecked")
+        IndexTombstone that = (IndexTombstone) other;
+        if (indexUUID.equals(that.indexUUID) == false) {
+            return false;
+        }
+        if (indexName.equals(that.indexName) == false) {
+            return false;
+        }
+        if (deleteDate != that.deleteDate) {
+            return false;
+        }
+        if (clusterVersion != that.clusterVersion) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(indexUUID, indexName, deleteDate, clusterVersion);
+    }
+
+    @Override
+    public String toString() {
+        return "IndexTombstone[indexUUID=" + indexUUID + ", indexName=" + indexName +
+               ", deleteDate=" + deleteDate + ", clusterVersion=" + clusterVersion + "]";
+    }
+
+    @Override
+    public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startObject();
+        builder.field(INDEX_UUID_KEY, indexUUID);
+        builder.field(INDEX_NAME_KEY, indexName);
+        builder.field(DELETE_DATE_KEY, deleteDate);
+        builder.field(CLUSTER_VERSION_KEY, clusterVersion);
+        builder.endObject();
+        return builder;
+    }
+
+    public static IndexTombstone fromXContent(final XContentParser parser) throws IOException {
+        return TOMBSTONE_PARSER.parse(parser, new IndexTombstone.Builder()).build();
+    }
+
+    /** A builder for building IndexTombstone objects. */
+    public static class Builder {
+        private String indexUUID;
+        private String indexName;
+        private long deleteDate = -1L;
+        private long clusterVersion = -1L;
+
+        public Builder() { }
+
+        public Builder indexUUID(final String indexUUID) {
+            this.indexUUID = indexUUID;
+            return this;
+        }
+
+        public String indexUUID() {
+            return indexUUID;
+        }
+
+        public Builder indexName(final String indexName) {
+            this.indexName = indexName;
+            return this;
+        }
+
+        public String indexName() {
+            return indexName;
+        }
+
+        public Builder deleteDate(final long deleteDate) {
+            this.deleteDate = deleteDate;
+            return this;
+        }
+
+        public long deleteDate() {
+            return deleteDate;
+        }
+
+        public Builder clusterVersion(final long clusterVersion) {
+            this.clusterVersion = clusterVersion;
+            return this;
+        }
+
+        public long clusterVersion() {
+            return clusterVersion;
+        }
+
+        public IndexTombstone build() {
+            assert indexUUID != null : "indexUUID must be set";
```

Comment:
these should all be real checks in the `ew IndexTombstone(indexUUID, indexName, deleteDate, clusterVersion);` ctor